### PR TITLE
Add Vector test support to Tril

### PR DESCRIPTION
**Pull request by mgaudet**
Allow writing tests using some vector constructs in Tril

The changes allow one to write a tril test case such as 

    (method return="NoType" args=[Address,Address,Address] 
     (block 
      (vstorei type=VectorDouble offset=0 
       (aload parm=0) 
       (vadd 
        (vloadi type=VectorDouble (aload parm=0))
        (vloadi type=VectorDouble (aload parm=1))))
       (return)))

### CODE REVIEW

**Comment by charliegracie** [comment]
genie-omr build all

File: fvtest/compilertriltest/VectorTest.cpp [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+TEST_F(VectorTest, VDoubleAdd) { 
+
+   auto inputTrees = "(method return= NoType args=[Address,Address,Address]           "
+                     "  (block                                                        "
+                     "     (vstorei type=VectorDouble offset=0                        "
+                     "         (aload parm=0)                                         "
+                     "            (vadd                                               "
+                     "                 (vloadi type=VectorDouble (aload parm=1))      "
+                     "                 (vloadi type=VectorDouble (aload parm=2))))    "
+                     "        (return)))                                              ";
```
Comment (Leonardo2718) [inline fvtest/compilertriltest/VectorTest.cpp]: I think the `(return)` should line up with the `(vstorei type=VectorDouble offset=0` line above (for readability).

File: fvtest/tril/tril/ilgen.cpp [hunk 2]
```diff
@@ -184,11 +185,34 @@ TR::Node* Tril::TRLangBuilder::toTRNode(const ASTNode* const tree) {
         }
      }
      else if (opcode.isLoadIndirect() || opcode.isStoreIndirect()) {
-         auto offset = tree->getArgByName("offset")->getValue()->get<int32_t>();
+         // If not specified, offset will default to zero. 
+         int32_t offset = 0; 
+         if (tree->getArgByName("offset")) {
+            offset = tree->getArgByName("offset")->getValue()->get<int32_t>();
+         } else { 
+            // warning("Should specify offset on indirect load and stores -- has defaulted to zero");
```
Comment (Leonardo2718) [inline fvtest/tril/tril/ilgen.cpp]: Maybe put a trace message here? Just to signal that the offset has defaulted to 0.

File: fvtest/tril/tril/ilgen.cpp [hunk 3]
```diff
@@ -184,11 +185,34 @@ TR::Node* Tril::TRLangBuilder::toTRNode(const ASTNode* const tree) {
...
+         } else { 
+            // warning("Should specify offset on indirect load and stores -- has defaulted to zero"); 
+         }
+
          TraceIL("  is indirect store/load with offset %d\n", offset);
          const auto name = tree->getName();
-         auto type = opcode.getType();
          auto compilation = TR::comp();
+         TR::DataType type;  
+         if (opcode.isVector()) {
```
Comment (Leonardo2718) [inline fvtest/tril/tril/ilgen.cpp]: Please add a trace message here to signal that we've detected a vector opcode.

File: fvtest/compilertriltest/VectorTest.cpp (hunk 1 above)
Comment (mgaudet) [reply fvtest/compilertriltest/VectorTest.cpp]: Fixed.

File: fvtest/tril/tril/ilgen.cpp (hunk 2 above)
Comment (mgaudet) [reply fvtest/tril/tril/ilgen.cpp]: Done -- I've reworked the tracing in that area as part of 751134f; 

output now looks like 

    Creating vstorei from ASTNode 0x7fbd59402c00
      is indirect store/load with offset 0  of vector type VectorDouble

File: fvtest/tril/tril/ilgen.cpp (hunk 3 above)
Comment (mgaudet) [reply fvtest/tril/tril/ilgen.cpp]: See earlier reply about trace.

**Comment by mgaudet** [comment]
There's a follow on task here to be done, once this is merged, to add support for the new vector type annotation to the `TrilDumper`.

**Comment by 0xdaryl** [comment]
genie-omr build all

File: fvtest/compilertriltest/CMakeLists.txt [hunk 4]
```diff
@@ -39,6 +39,7 @@ add_executable(comptest
 	ShiftAndRotateTest.cpp
 	SimplifierFoldAndTest.cpp
 	IfxcmpgeReductionTest.cpp
+    VectorTest.cpp
```
Comment (0xdaryl) [inline fvtest/compilertriltest/CMakeLists.txt]: Fix formatting, unless there's a reason why this is not aligned with the rest?

File: fvtest/tril/tril/ilgen.cpp [hunk 5]
```diff
@@ -184,11 +185,37 @@ TR::Node* Tril::TRLangBuilder::toTRNode(const ASTNode* const tree) {
...
          auto compilation = TR::comp();
+         TR::DataType type;  
+         if (opcode.isVector()) { 
+            // Vector types in TR IL are "typeless", insofar as they are
+            // supposed to infer the vector type depending on the children.
+            // Loads determine their data type based on the symref. However,
+            // given that we are creating a symref here, we need a hint as to
+            // what type of symref to create. So, vloadi and vstorei will take 
+            // an extra argument "type" to annotate the type desired.
+            if (tree->getArgByName("type") != nullptr) {
```
Comment (0xdaryl) [inline fvtest/tril/tril/ilgen.cpp]: Shouldn't this `nullptr` be `NULL` as required by #1718 ?

File: fvtest/tril/tril/ilgen.cpp [hunk 6]
```diff
@@ -184,11 +185,37 @@ TR::Node* Tril::TRLangBuilder::toTRNode(const ASTNode* const tree) {
...
+            // Loads determine their data type based on the symref. However,
+            // given that we are creating a symref here, we need a hint as to
+            // what type of symref to create. So, vloadi and vstorei will take 
+            // an extra argument "type" to annotate the type desired.
+            if (tree->getArgByName("type") != nullptr) { 
+               auto nameoftype = tree->getArgByName("type")->getValue()->getString();
+               type = getTRDataTypes(nameoftype); 
+               TraceIL(" of vector type %s\n", nameoftype);
+            } else { 
+               return nullptr;
```
Comment (0xdaryl) [inline fvtest/tril/tril/ilgen.cpp]: Same comment as above about `nullptr`.

File: fvtest/tril/tril/ilgen.cpp (hunk 5 above)
Comment (mgaudet) [reply fvtest/tril/tril/ilgen.cpp]: I had been matching the style of the file as it was, but you're right, this is going to make merging easier. Fixed. Ditto below.

File: fvtest/compilertriltest/CMakeLists.txt (hunk 4 above)
Comment (mgaudet) [reply fvtest/compilertriltest/CMakeLists.txt]: Fixed. Some spaces snuck in.

**Comment by 0xdaryl** [comment]
genie-omr build all

**Pull request merged**